src/JWTVerifier.php

```diff
@@ -5,160 +5,294 @@
-            }
+        if (! $this->supportsAlg($head_decoded->alg)) {
+            throw new InvalidTokenException('Token algorithm not supported.');
+        }
 
-            $jwks_url = $body->iss.$this->jwks_path;
-            $secret   = $this->JWKFetcher->fetchKeys($jwks_url);
-        } else if ($head->alg === 'HS256') {
-            if ($this->secret_base64_encoded) {
-                $secret = JWT::urlsafeB64Decode($this->client_secret);
-            } else {
-                $secret = $this->client_secret;
-            }
-        } else {
-            throw new InvalidTokenException('Invalid signature algorithm');
+        if (! is_object($body_decoded)) {
+            throw new InvalidTokenException('Malformed token payload.');
         }
 
-        try {
-            // Decode the user
-            $decodedToken = JWT::decode($jwt, $secret, ['HS256', 'RS256']);
-            // validate that this JWT was made for us
-            $audience = $decodedToken->aud;
-            if (! is_array($audience)) {
-                $audience = [$audience];
+        if (empty($body_decoded->aud)) {
+            throw new InvalidTokenException('Token audience not found.');
+        }
+
+        // Validate the token audience.
+        $audience = $body_decoded->aud;
+        if (! is_array($audience)) {
+            $audience = [$audience];
+        };
+        if (! count(array_intersect($audience, $this->valid_audiences))) {
+            throw new InvalidTokenException('This token is not intended for us.');
+        }
+
+        if ('HS256' === $head_decoded->alg) {
+            $secret = $this->client_secret;
+        } else {
+            if (empty($head_decoded->kid)) {
+                throw new CoreException('Token key ID is missing for RS256 token.');
             }
 
-            if (count(array_intersect($audience, $this->valid_audiences)) == 0) {
-                throw new InvalidTokenException('This token is not intended for us.');
+            if (empty($body_decoded->iss) || ! in_array($body_decoded->iss, $this->authorized_iss)) {
+                throw new CoreException('We cannot trust on a token issued by `'.$body_decoded->iss.'`.');
             }
+
+            $jwks_url                   = $body_decoded->iss.$this->jwks_path;
+            $secret[$head_decoded->kid] = $this->JWKFetcher->requestJwkX5c($jwks_url, $head_decoded->kid);
+        }
+
+        try {
+            return $this->decodeToken($jwt, $secret);
         } catch (\Exception $e) {
             throw new CoreException($e->getMessage());
         }
+    }
+
+    /**
+     * Wrapper for JWT::decode().
+     *
+     * @param string       $jwt    JWT to decode.
+     * @param string|array $secret Secret to use.
+     *
+     * @return mixed
+     *
+     * @codeCoverageIgnore
+     */
+    protected function decodeToken($jwt, $secret)
+    {
+        return JWT::decode($jwt, $secret, $this->supported_algs);
```

Comment:
Any reason why you moved this one-liner call to a separate method? If negative, consider calling it directly on the one place it's being used. If positive, consider making this private as people should be calling `verifyAndDecode` instead. (This method is also verifying the signature AFAIK, so the name is misleading 😛 )

Reply:
It was the only way to mock it, can't mock static methods so I couldn't just DI the JWT lib.